src/analyses/loopfreeCallstring.ml

```diff
@@ -1,60 +1,54 @@
 *)
 open Analyses
 
 module Spec : MCPSpec =
 struct
-  include Analyses.IdentitySpec
+  include UnitAnalysis.Spec
 
   let name () = "loopfree_callstring"
 
   module FundecSet = SetDomain.Make (CilType.Fundec)
-  module Either =  Printable.Either (CilType.Fundec) (FundecSet) 
-
-  module D = Lattice.Flat (Printable.Liszt (Either)) (* A domain element is a list containing fundecs and sets of fundecs.*) 
-  module C = D
-  module V = EmptyV
-  module G = Lattice.Unit
-  let startstate v = `Lifted([])
-  let exitstate  v = `Lifted([])
-
-  let get_list list = match list with
-    | `Lifted e -> e
-    | _ -> failwith "Error loopfreeCallstring (get_list): The list cannot be derived from Top or Bottom!"
-
-  let loop_detected f = function
-    (* note: a call string contains each Fundec at most once *)
-    | `Left ele -> CilType.Fundec.equal f ele
-    | `Right set -> FundecSet.mem f set 
-
-  let add_to_set old = function
-    | `Left ele -> FundecSet.add ele old
-    | `Right set -> FundecSet.join old set
-
-  let rec callee_state f prev_set prev_list = function
-    | [] -> (`Left f)::(List.rev prev_list) (* f is not yet contained in the call string *)
-    | e::rem_list -> 
-      let new_set = add_to_set prev_set e in
-      if loop_detected f e (* f is already present in the call string *)
-      then (`Right new_set)::rem_list (* combine all elements of the call cycle in a set *)
-      else callee_state f new_set (e::prev_list) rem_list
-
-  let callee_state f ctx = `Lifted(callee_state f (FundecSet.empty ()) [] (get_list ctx.local))
-
-  let enter ctx r f args = [ctx.local, callee_state f ctx]
-
-  let threadenter ctx ~multiple lval v args = [callee_state (Cilfacade.find_varinfo_fundec v) ctx]
-
-  let combine_env ctx lval fexp f args fc au f_ask = ctx.local
+  module Either =  Printable.Either (CilType.Fundec) (FundecSet)
+
+  module C = Printable.Liszt (Either)
+
+  let append fd current =
+    let loop_detected f = function
+      (* note: a call string contains each Fundec at most once *)
+      | `Left ele -> CilType.Fundec.equal f ele
+      | `Right set -> FundecSet.mem f set
+    in
+    let add_to_set old = function
+      | `Left ele -> FundecSet.add ele old
+      | `Right set -> FundecSet.join old set
+    in
+    let rec append f prev_set prev_list = function
+      | [] -> (`Left f)::(List.rev prev_list) (* f is not yet contained in the call string *)
+      | e::rem_list ->
+        let new_set = add_to_set prev_set e in
+        if loop_detected f e (* f is already present in the call string *)
+        then (`Right new_set)::rem_list (* combine all elements of the call cycle in a set *)
+        else append f new_set (e::prev_list) rem_list
+    in
+    append fd (FundecSet.empty ()) [] current
+
+  let context ctx fd x =
+    let curr_context =
+      try
+        ctx.context ()
+      with Enter_func_has_no_context -> []
```

Comment:
When one would (hypothetically) configure this analysis to be context-insensitive, would it crash with a `ctx_failure` exception? If so, that may even be desired behavior, but may be confusing to a user if the error message is unclear.